src/dividat-driver/firmware/main.go

```diff
@@ -1,134 +1,97 @@
+			return err
 		}
-	}
 
-	// 2: Switch the Senso to bootloader mode
-	if controllerHost != "" {
-		err := sendDfuCommand(controllerHost, controllerPort)
+		backoffStrategy := backoff.NewExponentialBackOff()
+		backoffStrategy.MaxElapsedTime = 30 * time.Second
+		backoffStrategy.MaxInterval = 10 * time.Second
+		err := backoff.RetryNotify(trySendDfu, backoffStrategy, func(e error, d time.Duration) {
+			onProgress(fmt.Sprintf("%v\nRetrying in %v", e, d))
+		})
+
 		if err != nil {
-			// Log the failure, but continue anyway, as the Senso might have been left in
-			// bootloader mode when a previous update process failed. Not all versions of
-			// the firmware automtically exit from the bootloader mode upon restart.
-			fmt.Printf("Could not send DFU command to Senso at %s: %s\n", controllerHost, err)
+			return fmt.Errorf("could not send DFU command to Senso at %s: %s", target.Address, err)
 		}
-	} else {
-		fmt.Printf("Could not discover a Senso in regular mode, now trying to detect a Senso already in bootloader mode.\n")
-	}
 
-	// 3: Find address of Senso in bootloader mode
-	var dfuHost string
-	if *configuredAddr != "" {
-		dfuHost = *configuredAddr
-	} else {
-		ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
-		discoveredAddr, err := discover("_sensoUpdate._udp", deviceSerial, ctx)
-		cancel()
-		if err != nil {
-			// Up to firmware 2.0.0.0 the bootloader advertised itself with the same
-			// service identifier as the application level firmware. To support such
-			// legacy devices, we look for `_sensoControl` again at this point, if
-			// the other service has not been found.
-			// We do need to rediscover, as the legacy device may still just have
-			// restarted into the bootloader and obtained a new IP address.
-			ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
-			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
-			cancel()
-			if err == nil {
-				dfuHost = legacyDiscoveredAddr
-			} else if controllerHost != "" {
-				fmt.Printf("Could not discover update service, trying to fall back to previous discovery %s.\n", controllerHost)
-				dfuHost = controllerHost
-			} else {
-				fail = fmt.Errorf("Could not find any Senso bootloader to transmit firmware to: %s", err)
-				return
-			}
-		} else {
-			dfuHost = discoveredAddr
+		onProgress("Looking for senso in bootloader mode")
+		dfuService := service.Find(parentCtx, 30*time.Second, func(discovered service.Service) bool {
+			return service.SerialNumberFilter(target.Text.Serial)(discovered) && service.IsDfuService(discovered)
+		})
+
+		if dfuService == nil {
+			return fmt.Errorf("Could not rediscover Senso in bootloader mode.\n%s", tryPowerCycling)
 		}
+
+		target = *dfuService
+		onProgress(fmt.Sprintf("Re-discovered Senso in bootloader mode at %s", target.Address))
+		onProgress("Waiting 15 seconds to ensure proper TFTP startup")
+		// Wait to ensure proper TFTP startup
+		time.Sleep(5 * time.Second)
```

Comment:
5 or 15? Is a long/any wait even still needed with the retry-logic?

Reply:
The wait is still necessary even with retry-logic. I tried removing the timeout, but while testing consecutive updates, after a few runs `putTFTP` fails. Not only that, but the Senso doesn't respond anymore when running a new update, and it needs to be power cycled or sometimes even a re-flashed. I've found 5 seconds to work quite well, but I'm still observing occasional failure after many runs. I'm not sure if it's because this timeout is still too short. I'm experimenting with longer timeouts and see if it has any influence.

I'm not sure why this happens. It seems like when the Senso receives a TFTP write request too early after rebooting into bootloader mode, it gets into a weird state that it cannot recover from.

Reply:
Increasing the wait to 10 seconds improved robustness. I was able to do over 250 back-to-back updates before killing the test script.

[Increase wait time before attempting TFTP transfer](https://github.com/dividat/driver/pull/127/commits/ffa7b584a1a33fedbb92dad271044cc829518aee)

Reply:
:+1: 

23 runs until cancelled by me:

    successful_runs=0; while bin/dividat-driver update-firmware -i ../senso-firmware/build/images/controller-app.bin; do ((successful_runs++)); echo "Runs: $successful_runs"; sleep 1; done

Good to go with this